contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,308 @@
+     * @param _denominated the token the result is denominated in. Must be the same as short.collateral for now.
+     * @return netValue the amount by which the margin is above or below the required amount.
+     * @return isExcess true if there's excess margin in the vault. In this case, collateral can be taken out from the vault. False if there is insufficient margin and additional collateral needs to be added to the vault to create the position.
+     */
+    function getExcessMargin(MarginAccount.Vault memory _vault, address _denominated)
+        public
+        view
+        returns (uint256 netValue, bool isExcess)
+    {
+        // ensure the number of collateral, long and short array is valid.
+        _checkIsValidSpread(_vault);
+        // ensure the long asset is valid for the short asset.
+        _checkIsMarginableLong(_vault);
+
+        // collateral amount is always positive.
+        FixedPointInt256.FixedPointInt memory collateralAmount = _vault.collateralAmounts.length > 0
+            ? _uint256ToFixedPointInt(_vault.collateralAmounts[0])
+            : _uint256ToFixedPointInt(0);
+
+        // Vault contains no short tokens: return collateral value.
+        if (_vault.shortOtokens.length == 0) return (SignedConverter.intToUint(collateralAmount.value), true);
+
+        // For the currenct version, ensure denominated == short.collateral
+        address shortCollateral = OtokenInterface(_vault.shortOtokens[0]).collateralAsset();
+        require(shortCollateral == _denominated, "MarginCalculator: Denomintated token should be short.collateral");
+
+        FixedPointInt256.FixedPointInt memory marginRequirement = _getMarginRequired(_vault);
+        // if marginRequirement > 0, the long assets cannot cover the max loss of short assets in the vault.
+        // will need to check if collateral - marginRequirement is greater than 0.
+        FixedPointInt256.FixedPointInt memory excessMargin = collateralAmount.sub(marginRequirement);
+        isExcess = excessMargin.isGreaterThanOrEqual(_uint256ToFixedPointInt(0));
+        netValue = SignedConverter.intToUint(excessMargin.value);
+    }
+
+    /**
+     * @notice Calculate the amount of collateral needed for a spread vault.
+     * @dev The vault passed in already pass amount array length = asset array length check.
+     * @param _vault the theoretical vault that needs to be checked
+     * @return marginRequired the minimal amount of collateral needed in a vault.
+     */
+    function _getMarginRequired(MarginAccount.Vault memory _vault)
+        internal
+        view
+        returns (FixedPointInt256.FixedPointInt memory marginRequired)
+    {
+        // The vault passed in has a short array == 1, so we can just use shortAmounts[0]
+        FixedPointInt256.FixedPointInt memory shortAmount = _uint256ToFixedPointInt(_vault.shortAmounts[0]);
+
+        bool hasLongInVault = _vault.longOtokens.length > 0;
```

Comment:
should be `bool hasLongInVault = (_vault.longOtokens.length > 0) && (_vault.longOtokens.[0] != address(0));`